### PR TITLE
feat(SFT-88): Utilize Craft's queue for handling Email/Integration jobs

### PR DESCRIPTION
**Pull request by seandelaney**
To allow jobs to still access any form field values when processed in the queue, I had to change the following. Otherwise the values are blank.

From:
```
$fields = $form->getLayout()->getFields();
```

to: 

```
$fields = $submission->getFieldCollection();
```

Same applies for this next change. If we use fields collection to grab the field, then our value exists. Otherwise its blank:

From:
```
$field = $form->get($condition->getField()->getId());
$postedValue = $field?->getValue();
```

to:
```
$field = $fields->get($condition->getField());
if (!$field) {
    continue;
}

$postedValue = $field->getValue();
```

We now only handle integrations if there are integrations setup. Hopefully help performance wise.

### CODE REVIEW

**Review by kjmartens — CHANGES_REQUESTED** [review]
I am immediately running into a couple of issues that seem to relate to the Google Sheets integration. I have the integration configured, but I don't have it enabled for all my forms. For forms that do NOT use it, the submit still fails and includes and error like this:

```
<pre>Exception 'TypeError' with message 'min(): Argument #1 ($value) must be of type array, int given' 

in /Users/kmartens/Repos/craft-freeform/packages/plugin/src/Integrations/Other/Google/GoogleSheets/Versions/GoogleSheetsV4.php:83

Stack trace:
#0 /Users/kmartens/Repos/craft-freeform/packages/plugin/src/Integrations/Other/Google/GoogleSheets/Versions/GoogleSheetsV4.php(83): min(9999999)
#1 /Users/kmartens/Repos/craft-freeform/packages/plugin/src/Services/Integrations/IntegrationsService.php(478): Solspace\Freeform\Integrations\Other\Google\GoogleSheets\Versions\GoogleSheetsV4-&gt;push(Object(Solspace\Freeform\Form\Types\Regular), Object(GuzzleHttp\Client))
#2 /Users/kmartens/Repos/craft-freeform/packages/plugin/src/Integrations/Other/Google/GoogleSheets/GoogleSheetsBundle.php(56): Solspace\Freeform\Services\Integrations\IntegrationsService-&gt;processIntegrations(Object(Solspace\Freeform\Form\Types\Regular), 'Solspace\\Freefo...')
#3 [internal function]: Solspace\Freeform\Integrations\Other\Google\GoogleSheets\GoogleSheetsBundle-&gt;pushToSheets(Object(Solspace\Freeform\Events\Submissions\ProcessSubmissionEvent))
#4 /Users/kmartens/Herd/craft5upgrade/vendor/yiisoft/yii2/base/Event.php(312): call_user_func(Array, Object(Solspace\Freeform\Events\Submissions\ProcessSubmissionEvent))
#5 /Users/kmartens/Repos/craft-freeform/packages/plugin/src/Services/SubmissionsService.php(200): yii\base\Event::trigger('Solspace\\Freefo...', 'process-submiss...', Object(Solspace\Freeform\Events\Submissions\ProcessSubmissionEvent))
#6 /Users/kmartens/Repos/craft-freeform/packages/plugin/src/Services/SubmissionsService.php(170): Solspace\Freeform\Services\SubmissionsService-&gt;postProcessSubmission(Object(Solspace\Freeform\Form\Types\Regular), Object(Solspace\Freeform\Elements\Submission))
#7 /Users/kmartens/Repos/craft-freeform/packages/plugin/src/controllers/SubmitController.php(46): Solspace\Freeform\Services\SubmissionsService-&gt;handleSubmission(Object(Solspace\Freeform\Form\Types\Regular))
#8 [internal function]: Solspace\Freeform\controllers\SubmitController-&gt;actionIndex()
#9 /Users/kmartens/Herd/craft5upgrade/vendor/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array(Array, Array)
#10 /Users/kmartens/Herd/craft5upgrade/vendor/yiisoft/yii2/base/Controller.php(178): yii\base\InlineAction-&gt;runWithParams(Array)
#11 /Users/kmartens/Herd/craft5upgrade/vendor/yiisoft/yii2/base/Module.php(552): yii\base\Controller-&gt;runAction('', Array)
#12 /Users/kmartens/Herd/craft5upgrade/vendor/craftcms/cms/src/web/Application.php(349): yii\base\Module-&gt;runAction('freeform/submit', Array)
#13 /Users/kmartens/Herd/craft5upgrade/vendor/craftcms/cms/src/web/Application.php(650): craft\web\Application-&gt;runAction('freeform/submit', Array)
#14 /Users/kmartens/Herd/craft5upgrade/vendor/craftcms/cms/src/web/Application.php(311): craft\web\Application-&gt;_processActionRequest(Object(craft\web\Request))
#15 /Users/kmartens/Herd/craft5upgrade/vendor/yiisoft/yii2/base/Application.php(384): craft\web\Application-&gt;handleRequest(Object(craft\web\Request))
#16 /Users/kmartens/Herd/craft5upgrade/web/index.php(12): yii\base\Application-&gt;run()
#17 /Applications/Herd.app/Contents/Resources/valet/server.php(153): require('/Users/kmartens...')
#18 {main}</pre>
```

Or this:

```
<pre>Exception 'Solspace\Freeform\Library\Exceptions\Integrations\IntegrationException' with message 'Some or all of the configuration values are missing' 

in /Users/kmartens/Repos/craft-freeform/packages/plugin/src/Bundles/Integrations/OAuth/OAuth2RefreshTokenBundle.php:117

Stack trace:
#0 [internal function]: Solspace\Freeform\Bundles\Integrations\OAuth\OAuth2RefreshTokenBundle-&gt;refreshTokens(Object(Solspace\Freeform\Events\Integrations\GetAuthorizedClientEvent))
#1 /Users/kmartens/Herd/craft5upgrade/vendor/yiisoft/yii2/base/Event.php(312): call_user_func(Array, Object(Solspace\Freeform\Events\Integrations\GetAuthorizedClientEvent))
#2 /Users/kmartens/Repos/craft-freeform/packages/plugin/src/Bundles/Integrations/Providers/IntegrationClientProvider.php(17): yii\base\Event::trigger('Solspace\\Freefo...', 'get-client', Object(Solspace\Freeform\Events\Integrations\GetAuthorizedClientEvent))
#3 /Users/kmartens/Repos/craft-freeform/packages/plugin/src/Services/Integrations/IntegrationsService.php(477): Solspace\Freeform\Bundles\Integrations\Providers\IntegrationClientProvider-&gt;getAuthorizedClient(Object(Solspace\Freeform\Integrations\Other\Google\GoogleSheets\Versions\GoogleSheetsV4))
#4 /Users/kmartens/Repos/craft-freeform/packages/plugin/src/Integrations/Other/Google/GoogleSheets/GoogleSheetsBundle.php(56): Solspace\Freeform\Services\Integrations\IntegrationsService-&gt;processIntegrations(Object(Solspace\Freeform\Form\Types\Regular), 'Solspace\\Freefo...')
#5 [internal function]: Solspace\Freeform\Integrations\Other\Google\GoogleSheets\GoogleSheetsBundle-&gt;pushToSheets(Object(Solspace\Freeform\Events\Submissions\ProcessSubmissionEvent))
#6 /Users/kmartens/Herd/craft5upgrade/vendor/yiisoft/yii2/base/Event.php(312): call_user_func(Array, Object(Solspace\Freeform\Events\Submissions\ProcessSubmissionEvent))
#7 /Users/kmartens/Repos/craft-freeform/packages/plugin/src/Services/SubmissionsService.php(200): yii\base\Event::trigger('Solspace\\Freefo...', 'process-submiss...', Object(Solspace\Freeform\Events\Submissions\ProcessSubmissionEvent))
#8 /Users/kmartens/Repos/craft-freeform/packages/plugin/src/Services/SubmissionsService.php(170): Solspace\Freeform\Services\SubmissionsService-&gt;postProcessSubmission(Object(Solspace\Freeform\Form\Types\Regular), Object(Solspace\Freeform\Elements\Submission))
#9 /Users/kmartens/Repos/craft-freeform/packages/plugin/src/controllers/SubmitController.php(46): Solspace\Freeform\Services\SubmissionsService-&gt;handleSubmission(Object(Solspace\Freeform\Form\Types\Regular))
#10 [internal function]: Solspace\Freeform\controllers\SubmitController-&gt;actionIndex()
#11 /Users/kmartens/Herd/craft5upgrade/vendor/yiisoft/yii2/base/InlineAction.php(57): call_user_func_array(Array, Array)
#12 /Users/kmartens/Herd/craft5upgrade/vendor/yiisoft/yii2/base/Controller.php(178): yii\base\InlineAction-&gt;runWithParams(Array)
#13 /Users/kmartens/Herd/craft5upgrade/vendor/yiisoft/yii2/base/Module.php(552): yii\base\Controller-&gt;runAction('', Array)
#14 /Users/kmartens/Herd/craft5upgrade/vendor/craftcms/cms/src/web/Application.php(349): yii\base\Module-&gt;runAction('freeform/submit', Array)
#15 /Users/kmartens/Herd/craft5upgrade/vendor/craftcms/cms/src/web/Application.php(650): craft\web\Application-&gt;runAction('freeform/submit', Array)
#16 /Users/kmartens/Herd/craft5upgrade/vendor/craftcms/cms/src/web/Application.php(311): craft\web\Application-&gt;_processActionRequest(Object(craft\web\Request))
#17 /Users/kmartens/Herd/craft5upgrade/vendor/yiisoft/yii2/base/Application.php(384): craft\web\Application-&gt;handleRequest(Object(craft\web\Request))
#18 /Users/kmartens/Herd/craft5upgrade/web/index.php(12): yii\base\Application-&gt;run()
#19 /Applications/Herd.app/Contents/Resources/valet/server.php(153): require('/Users/kmartens...')
#20 {main}</pre>
```

**Review by kjmartens — CHANGES_REQUESTED** [review]
When using the Queue, email notifications appear to work correctly. However, integrations do not.

I created 4 test submissions with an Admin email and then Mailchimp or Salesforce integration. They all queued up, but when they ran together after visiting CP, the email notifications came in, and the Mailchimp contacts and Salesforce Leads did NOT.

Without the queue, the Email and integrations work fine.

While the Mailchimp integration seems to just have nothing happen, the Salesforce integration jobs actually fail with the following error (again, the fields ARE mapped, and it works fine when not using Queue):

```
Client error: `POST https://d61000000zofvea0-dev-ed.my.salesforce.com/services/data/v58.0/sobjects/Lead` resulted in a `400 Bad Request` response: [{"message":"Required fields are missing: [LastName, Company]","errorCode":"REQUIRED_FIELD_MISSING","fields":["LastName" (truncated...)
```

Screenshot 2024-06-07 at 11 20 29 AM

**Comment by gustavs-gutmanis** [comment]
@kjmartens please stop merging other branches into PR branches, instead use Rebase if you must, otherwise the file history becomes tainted and hard to review.

File: packages/plugin/src/Bundles/Notifications/SendListeners/ConditionalNotifications.php [hunk 1]
```diff
@@ -31,62 +41,23 @@ public static function isProOnly(): bool
...
-                ->sendEmail(
-                    $form,
-                    $recipients,
-                    $fields,
-                    $template,
-                    $submission,
-                )
-            ;
+        if (!$this->notificationsProvider->getByFormAndClass($form, Conditional::class)) {
+            return;
```
Comment (gustavs-gutmanis) [inline packages/plugin/src/Bundles/Notifications/SendListeners/ConditionalNotifications.php]: I would keep the recipient and template gathering here, and create a job with the recipients and template, form ID and submission ID.

The job should be generic, for all notifications. The only thing that differs between them are the recipients and template. And you can gather those before creating a job, and passing it to the job.

The job would then fetch form, hydrate it, create the `$fields` array, and call the mailer service `::sendMail()` method.

This way we get rid of all the custom jobs and we don't put business logic in there, that would remain in feature bundles instead

File: packages/plugin/src/Jobs/FreeformQueueHandler.php [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+        $queue = \Craft::$app->getQueue();
+
+        if ($this->settingsService->isIntegrationQueueEnabled()) {
+            $queue->push($job);
+        } else {
+            $job->execute($queue);
+        }
+    }
+
+    public function rehydrateForm(Form $form, Submission $submission): Form
```
Comment (gustavs-gutmanis) [inline packages/plugin/src/Jobs/FreeformQueueHandler.php]: We should move this to the `Form` class.
`$form->valuesFromSubmission(Submission $submission)`

File: packages/plugin/src/Jobs/IntegrationJobInterface.php [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+ * @see           https://docs.solspace.com/craft/freeform
+ *
+ * @license       https://docs.solspace.com/license-agreement
+ */
+
+namespace Solspace\Freeform\Jobs;
+
+interface IntegrationJobInterface
+{
+    public function execute($queue): void;
```
Comment (gustavs-gutmanis) [inline packages/plugin/src/Jobs/IntegrationJobInterface.php]: This is already defined in the `JobInterface`

`interface IntegrationJobInterface extends JobInterface` and remove the `execute()` function definition from here.

File: packages/plugin/src/Jobs/NotificationJobInterface.php [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+ * @see           https://docs.solspace.com/craft/freeform
+ *
+ * @license       https://docs.solspace.com/license-agreement
+ */
+
+namespace Solspace\Freeform\Jobs;
+
+interface NotificationJobInterface
+{
+    public function execute($queue): void;
```
Comment (gustavs-gutmanis) [inline packages/plugin/src/Jobs/NotificationJobInterface.php]: Same as for Integration job

File: packages/plugin/src/Bundles/Notifications/SendListeners/ConditionalNotifications.php (hunk 1 above)
Comment (seandelaney) [reply packages/plugin/src/Bundles/Notifications/SendListeners/ConditionalNotifications.php]: @gustavs-gutmanis I had the recipient and template passed into the Job at one point during my refactor y'day. But then moved away from this again after getting a `Exception 'Exception' with message 'Serialization of 'Closure' is not allowed'` error for the Conditional notification when testing. 

I will revisit this again to see if I can get this working as you want.

**Review by gustavs-gutmanis — APPROVED** [review]
Looking good, lets wait for Kelsey now

**Review by kjmartens — APPROVED** [review]
Works swimmingly now

**Pull request merged**